### PR TITLE
CocoaAction separation

### PR DESCRIPTION
**Pull request by RuiAAPeres**
Since we want a "pure" repo for the Swift implementation and CocoaAction uses KVO, it should be separated.

### CODE REVIEW

**Comment by RuiAAPeres** [comment]
Since you can't have a stored property in an extension, should we use associated properties for this? I could also aim this PR to the 5.0 branch and remove this from the API.

**Comment by RuiAAPeres** [comment]
The issue with the tests, is that since we are not holding strongly to the `CocoaAction`, it get's out of scope and the tests fail. This dramatically change the semantics, holding it strongly fixes the issue.

File: ReactiveCocoa/Swift/CocoaAction.swift [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+		_execute(input)
+	}
+	
+	public override class func automaticallyNotifiesObserversForKey(key: String) -> Bool {
+		return false
+	}
+}
+
+extension Action {
+
```
Comment (mdiep) [inline ReactiveCocoa/Swift/CocoaAction.swift]: This empty line should be removed to adhere to the RAC style guide.

**Comment by mdiep** [comment]
> We can't have a stored property in an extension, should we use associated properties for this?

Yeah, I think so. 👍 

Thanks for working on this! ⚡

**Comment by RuiAAPeres** [comment]
@mdiep What do you think about [this](https://github.com/neilpa/Rex/blob/master/Source/Foundation/Association.swift#L70#L77)? When we move Rex to this org, it could use that specific method, so we would remove duplication then.

**Comment by mdiep** [comment]
I think including that is a fine idea. 👍

**Comment by ikesyo** [comment]
This seems to be ready now.

File: ReactiveCocoa/Swift/Association.swift [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import Foundation
+
+/// On first use attaches the object returned from `initial` to the `host` object using
+/// `key` via `objc_setAssociatedObject`. On subsequent usage, returns said object via
+/// `objc_getAssociatedObject`.
+public func associatedObject<Host: AnyObject, T: AnyObject>(host: Host, key: UnsafePointer<()>, @noescape initial: Host -> T) -> T {
```
Comment (mdiep) [inline ReactiveCocoa/Swift/Association.swift]: I think we should make this `internal`.

**Comment by mdiep** [comment]
Just the one note. ☺️

**Comment by RuiAAPeres** [comment]
If we make it internal, we won't be able to use it once Rex is in this org (or even before that). Are you ok with this duplication?

**Comment by mdiep** [comment]
> If we make it internal, we won't be able to use it once Rex is in this org (or even before that). Are you ok with this duplication?

Yes. Although my thinking is that we should move the Rex code into this repo as part of the split. Let's make it `internal` for now and discuss the rest later. 😄

**Comment by RuiAAPeres** [comment]
> Although my thinking is that we should move the Rex code into this repo as part of the split.

You mean, adding Rex to the Swift + Foundation Extension?

**Comment by mdiep** [comment]
https://github.com/ReactiveCocoa/ReactiveCocoa/issues/2790#issuecomment-218942290 😄

**Comment by RuiAAPeres** [comment]
@mdiep I misunderstood that bit I thought:

1. Swift + Foundation
2. Rex (with a new name)

**Comment by mdiep** [comment]
✨ 

> I misunderstood that bit, I thought:
> 
> 1. Swift + Foundation
> 2. Rex (with a new name)

I'm just proposing that Rex's new name should be ReactiveCocoa and it should be merged with the things we currently have that don't fit with Swift or Obj-C.

**Pull request merged**